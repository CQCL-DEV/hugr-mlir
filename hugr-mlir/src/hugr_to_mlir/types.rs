use crate::mlir::hugr::ffi::mlirHugrTranslateStringRefToMLIRFunction;
use crate::{mlir, Error, Result};
use hugr::hugr::IdentList;
use melior::Context;
use std::borrow::Borrow;
use std::ops::Deref;

pub fn extension_id_to_extension_attr<'c>(
    context: &'c Context,
    id: &'_ hugr::extension::ExtensionId,
) -> mlir::hugr::ExtensionAttr<'c> {
    crate::mlir::hugr::ExtensionAttr::new(context, id.borrow())
}

pub fn extension_set_to_extension_set_attr<'c>(
    context: &'c Context,
    set: &'_ hugr::extension::ExtensionSet,
) -> mlir::hugr::ExtensionSetAttr<'c> {
    crate::mlir::hugr::ExtensionSetAttr::new(
        context,
        set.iter()
            .map(|x| extension_id_to_extension_attr(context, x))
            .collect::<Vec<_>>(),
    )
}

pub fn hugr_to_mlir_type<'c>(
    ctx: &'c Context,
    type_: &hugr::types::Type,
) -> Result<melior::ir::Type<'c>, Error> {
    use hugr::types::TypeEnum;
    match type_.as_type_enum() {
        TypeEnum::Sum(sum_type) => {
            let mut alts = std::vec::Vec::<_>::new();
            for i in 0usize.. {
                if let Some(t) = sum_type.get_variant(i) {
                    alts.push(hugr_to_mlir_type(ctx, t)?)
                } else {
                    break;
                }
            }
            Ok(mlir::hugr::SumType::new(ctx, alts).into())
        }
        TypeEnum::Tuple(row) => Ok(melior::ir::r#type::TupleType::new(
            ctx,
            collect_type_row::<Vec<_>>(ctx, row)?.as_slice(),
        )
        .into()),
        TypeEnum::Extension(ref custom_type) => {
            let mb_t = match custom_type.extension().deref() {
                "prelude" => match custom_type.name().as_str() {
                    "usize" => Some(unsafe {
                        melior::ir::Type::from_raw(mlir_sys::mlirIndexTypeGet(ctx.to_raw()))
                    }),
                    _ => None
                },
                "arithmetic.int.types" => match custom_type.name().as_str() {
                    "int" => {
                        assert!(!custom_type.args().is_empty());
                        if let hugr::types::type_param::TypeArg::BoundedNat { n } =
                            custom_type.args()[0]
                        {
                            Some(melior::ir::r#type::IntegerType::new(ctx, n as u32).into())
                        } else {
                            panic!(
                                "int type does not have a bounded nat arg: {:?}",
                                custom_type
                            )
                        }
                    }
                    _ => None
                },
                "arithmetic.float.types" => match custom_type.name().as_str() {
                    "float64" => {
                        assert!(custom_type.args().is_empty());
                        Some(unsafe {
                            melior::ir::Type::from_raw(mlir_sys::mlirF64TypeGet(ctx.to_raw()))
                        })
                    }
                    "float32" => {
                        assert!(custom_type.args().is_empty());
                        Some(unsafe {
                            melior::ir::Type::from_raw(mlir_sys::mlirF32TypeGet(ctx.to_raw()))
                        })
                    }
                    _ => None
                },
                _ => None
            };
            if let Some(t) = mb_t {
                Ok(t)
            } else {
                Ok(crate::mlir::hugr::OpaqueType::new(
                    custom_type.name().as_ref(),
                    extension_id_to_extension_attr(ctx, custom_type.extension()),
                    collect_type_args::<Vec<_>>(ctx, custom_type.args())?,
                    mlir::hugr::TypeConstraintAttr::new(ctx, custom_type.bound()),
                    ).into())
            }
        },
        TypeEnum::Alias(ref alias_decl) => Ok(mlir::hugr::AliasRefType::new(
            mlir::hugr::ExtensionSetAttr::new(ctx, []),
            melior::ir::attribute::FlatSymbolRefAttribute::new(ctx, alias_decl.name.as_ref())
                .into(),
            mlir::hugr::TypeConstraintAttr::new(ctx, alias_decl.bound),
        )
        .into()),
        TypeEnum::Function(ref function_type) if function_type.params().is_empty() => {
            Ok(hugr_to_mlir_function_type(ctx, function_type.body())?.into())
        }
        TypeEnum::Function(ref _function_type) => {
            panic!("unimplemented: TypeEnum::Function with params")
        }
        TypeEnum::Variable(_size, _bound) => panic!("unimplemented: TypeEnum::Variable"),
    }
}

pub fn hugr_to_mlir_function_type<'a>(
    ctx: &'a Context,
    type_: &'_ hugr::types::FunctionType,
) -> Result<mlir::hugr::HugrFunctionType<'a>, Error> {
    let type_extensions = extension_set_to_extension_set_attr(ctx, &type_.extension_reqs);
    let type_inputs = collect_type_row::<Vec<_>>(ctx, &type_.input)?;
    let type_outputs = collect_type_row::<Vec<_>>(ctx, &type_.output)?;
    Ok(mlir::hugr::HugrFunctionType::new(
        type_extensions,
        melior::ir::r#type::FunctionType::new(ctx, type_inputs.as_slice(), type_outputs.as_slice()),
    ))
}

pub fn collect_type_row_vec<'c>(
    context: &'c Context,
    row: &'_ hugr::types::TypeRow,
) -> Result<Vec<melior::ir::Type<'c>>> {
    collect_type_row(context, row)
}

pub fn collect_type_row<'c, R: FromIterator<melior::ir::Type<'c>>>(
    context: &'c Context,
    row: &'_ hugr::types::TypeRow,
) -> Result<R, Error> {
    row.iter().map(|x| hugr_to_mlir_type(context, x)).collect()
}

pub fn collect_type_args<'c, 'a, R: FromIterator<melior::ir::Attribute<'c>>>(
    ctx: &'c Context,
    args: impl IntoIterator<Item = &'a hugr::types::TypeArg>,
) -> Result<R, Error> {
    use hugr::types::TypeArg;
    args.into_iter()
        .map(|x| match x {
            TypeArg::Type { ty } => {
                Ok(melior::ir::attribute::TypeAttribute::new(hugr_to_mlir_type(ctx, ty)?).into())
            }
            &TypeArg::BoundedNat { n } => Ok(melior::ir::attribute::IntegerAttribute::new(
                n as i64,
                melior::ir::r#type::IntegerType::new(ctx, 64).into(),
            )
            .into()),
            &TypeArg::Opaque { .. } => todo!(),
            TypeArg::Sequence { ref elems } => Ok(melior::ir::attribute::ArrayAttribute::new(
                ctx,
                &collect_type_args::<Vec<_>>(ctx, elems)?,
            )
            .into()),
            TypeArg::Extensions { es } => Ok(extension_set_to_extension_set_attr(ctx, es).into()),
            ta => panic!("unimplemented type_arg: {:?}", ta),
        })
        .collect::<Result<R, Error>>()
}
