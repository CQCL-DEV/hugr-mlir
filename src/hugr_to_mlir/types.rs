use std::borrow::Borrow;
use crate::{mlir, Error};

pub fn extension_id_to_extension_attr<'c>(
    context: &'c melior::Context,
    id: &'_ hugr::extension::ExtensionId,
) -> mlir::hugr::ExtensionAttr<'c> {
    crate::mlir::hugr::ExtensionAttr::new(context, id.borrow())
}

pub fn extension_set_to_extension_set_attr<'c>(
    context: &'c melior::Context,
    set: &'_ hugr::extension::ExtensionSet,
) -> mlir::hugr::ExtensionSetAttr<'c> {
    crate::mlir::hugr::ExtensionSetAttr::new(
        context,
        set.iter()
            .map(|x| extension_id_to_extension_attr(context, x))
            .collect::<Vec<_>>(),
    )
}

pub fn hugr_to_mlir_type<'c>(
    ctx: &'c melior::Context,
    type_: &'_ hugr::types::Type,
) -> Result<melior::ir::Type<'c>, Error> {
    use hugr::types::{PrimType, TypeEnum};
    match type_.clone().into() {
        TypeEnum::Sum(sum_type) => {
            let mut alts = std::vec::Vec::<_>::new();
            for i in 0usize.. {
                if let Some(t) = sum_type.get_variant(i) {
                    alts.push(hugr_to_mlir_type(ctx, t)?)
                } else {
                    break;
                }
            }
            Ok(mlir::hugr::SumType::new(ctx, alts).into())
        }
        TypeEnum::Tuple(row) => Ok(melior::ir::r#type::TupleType::new(
            ctx,
            collect_type_row::<Vec<_>>(ctx, &row)?.as_slice(),
        )
        .into()),
        TypeEnum::Prim(p) => match p {
            PrimType::Extension(custom_type) => Ok(crate::mlir::hugr::OpaqueType::new(
                custom_type.name().as_ref(),
                extension_id_to_extension_attr(ctx, custom_type.extension()),
                collect_type_args::<Vec<_>>(ctx, custom_type.args())?,
                mlir::hugr::TypeConstraintAttr::new(ctx, custom_type.bound()),
            )
            .into()),
            PrimType::Alias(alias_decl) => Ok(mlir::hugr::AliasRefType::new(
                mlir::hugr::ExtensionSetAttr::new(ctx, []),
                melior::ir::attribute::FlatSymbolRefAttribute::new(ctx, alias_decl.name.as_ref())
                    .into(),
                mlir::hugr::TypeConstraintAttr::new(ctx, alias_decl.bound),
            )
            .into()),
            PrimType::Function(function_type) => {
                Ok(hugr_to_mlir_function_type(ctx, &function_type)?.into())
            }
        },
    }
}

pub fn hugr_to_mlir_function_type<'a>(
    ctx: &'a melior::Context,
    type_: &'_ hugr::types::FunctionType,
) -> Result<mlir::hugr::HugrFunctionType<'a>, Error> {
    let type_extensions = extension_set_to_extension_set_attr(ctx, &type_.extension_reqs);
    let type_inputs = collect_type_row::<Vec<_>>(ctx, &type_.input)?;
    let type_outputs = collect_type_row::<Vec<_>>(ctx, &type_.output)?;
    Ok(mlir::hugr::HugrFunctionType::new(
        type_extensions,
        melior::ir::r#type::FunctionType::new(ctx, type_inputs.as_slice(), type_outputs.as_slice()),
    ))
}

pub fn collect_type_row<'c, R: FromIterator<melior::ir::Type<'c>>>(
    context: &'c melior::Context,
    row: &'_ hugr::types::TypeRow,
) -> Result<R, Error> {
    row.iter().map(|x| hugr_to_mlir_type(context, x)).collect()
}

pub fn collect_type_args<'c, 'a, R: FromIterator<melior::ir::Attribute<'c>>>(
    ctx: &'c melior::Context,
    args: impl IntoIterator<Item = &'a hugr::types::TypeArg>,
) -> Result<R, Error> {
    use hugr::types::TypeArg;
    args.into_iter()
        .map(|x| match x {
            &TypeArg::Type { ref ty } => {
                Ok(melior::ir::attribute::TypeAttribute::new(hugr_to_mlir_type(ctx, ty)?).into())
            }
            &TypeArg::BoundedNat { n } => Ok(melior::ir::attribute::IntegerAttribute::new(
                n as i64,
                melior::ir::r#type::IntegerType::new(ctx, 64).into(),
            )
            .into()),
            &TypeArg::Opaque { .. } => todo!(),
            &TypeArg::Sequence { ref args } => Ok(melior::ir::attribute::ArrayAttribute::new(
                ctx,
                &collect_type_args::<Vec<_>>(ctx, args)?,
            )
            .into()),
            &TypeArg::Extensions { ref es } => {
                Ok(extension_set_to_extension_set_attr(ctx, es).into())
            }
            ta => panic!("unimplemented type_arg: {:?}", ta),
        })
        .collect::<Result<R, Error>>()
}
