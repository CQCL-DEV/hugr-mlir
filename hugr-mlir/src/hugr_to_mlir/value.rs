use crate::{mlir, Error, Result};
use anyhow::anyhow;

use itertools::zip_eq;
use melior::Context;

use super::types::hugr_to_mlir_type;

pub fn hugr_to_mlir_value<'c>(
    context: &'c Context,
    typ: &hugr::types::Type,
    value: &hugr::values::Value,
) -> Result<melior::ir::Attribute<'c>> {
    use downcast_rs::Downcast;
    use hugr::types::TypeEnum;
    use hugr::values::Value;
    match value {
        Value::Tuple { vs } => {
            let TypeEnum::Tuple(ref typerow) = typ.as_type_enum() else { Err(anyhow!("not a tuple type"))? };
            Ok(melior::ir::attribute::ArrayAttribute::new(
                context,
                zip_eq(typerow.iter(), vs.iter())
                    .map(|(ty, x)| hugr_to_mlir_value(context, ty, x))
                    .collect::<Result<Vec<_>>>()?
                    .as_slice(),
            )
            .into())
        }
        &Value::Sum { tag, ref value } => {
            let TypeEnum::Sum(ref sum_type) =  typ.as_type_enum() else { Err(anyhow!("not a sum type"))? };
            let Some(variant_type) = sum_type.get_variant(tag) else { Err(anyhow!("bad tag for sum type"))? };
            Ok(mlir::hugr::SumAttribute::new(
                hugr_to_mlir_type(context, typ)?,
                tag as u32,
                hugr_to_mlir_value(context, variant_type, value)?,
            )
            .into())
        }
        &Value::Extension {ref c } => {
            if let Some(i) =
                c.0.downcast_ref::<hugr::std_extensions::arithmetic::int_types::ConstIntS>()
            {
                Ok(melior::ir::attribute::IntegerAttribute::new(
                    i.value(),
                    melior::ir::r#type::IntegerType::new(context, i.log_width().into()).into(),
                )
                .into())
            } else if let Some(i) =
                c.0.downcast_ref::<hugr::std_extensions::arithmetic::int_types::ConstIntU>()
            {
                // this as i64 is naughty
                Ok(melior::ir::attribute::IntegerAttribute::new(
                    i.value() as i64,
                    melior::ir::r#type::IntegerType::new(context, i.log_width().into()).into(),
                )
                .into())
            } else if let Some(f) =
                c.0.downcast_ref::<hugr::std_extensions::arithmetic::float_types::ConstF64>()
            {
                Ok(
                    melior::ir::attribute::FloatAttribute::new(context, f.value(), unsafe {
                        melior::ir::Type::from_raw(mlir_sys::mlirF64TypeGet(context.to_raw()))
                    })
                    .into(),
                )
            } else if let Some(f) = c.0.downcast_ref::<hugr::extension::prelude::ConstUsize>() {
                Ok(
                    melior::ir::attribute::IntegerAttribute::new(f.value() as i64, unsafe {
                        melior::ir::Type::from_raw(mlir_sys::mlirIndexTypeGet(context.to_raw()))
                    })
                    .into(),
                )
            } else {
                Err(anyhow!(
                    "hugr_to_mlir_value:unimplemented extension constant: {:?}",
                    c
                ))
            }
        }
        x => Err(anyhow!("Unimplemented hugr_to_mlir_value: {:?}", x)),
    }
}
