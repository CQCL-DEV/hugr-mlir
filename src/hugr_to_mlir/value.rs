use crate::{mlir, Error, Result};
use anyhow::anyhow;

use itertools::zip_eq;
use melior::Context;

use super::types::hugr_to_mlir_type;

pub fn hugr_to_mlir_value<'c>(
    context: &'c Context,
    typ: &hugr::types::Type,
    value: &hugr::values::Value,
) -> Result<melior::ir::Attribute<'c>> {
    use hugr::types::TypeEnum;
    use hugr::values::Value;
    match value {
        &Value::Tuple { ref vs } => {
            let TypeEnum::Tuple(ref typerow) = typ.as_type_enum() else { Err(anyhow!("not a tuple type"))? };
            Ok(melior::ir::attribute::ArrayAttribute::new(
                context,
                zip_eq(typerow.iter(), vs.iter())
                    .map(|(ty, x)| hugr_to_mlir_value(context, ty, x))
                    .collect::<Result<Vec<_>>>()?
                    .as_slice(),
            )
            .into())
        }
        &Value::Sum { tag, ref value } => {
            let TypeEnum::Sum(ref sum_type) =  typ.as_type_enum() else { Err(anyhow!("not a sum type"))? };
            let Some(variant_type) = sum_type.get_variant(tag) else { Err(anyhow!("bad tag for sum type"))? };
            Ok(mlir::hugr::SumAttribute::new(
                hugr_to_mlir_type(context, typ)?,
                tag as u32,
                hugr_to_mlir_value(context, variant_type, value)?,
            )
            .into())
        }
        x => Err(anyhow!("Unimplemented hugr_to_mlir_value: {:?}", x)),
    }
}
