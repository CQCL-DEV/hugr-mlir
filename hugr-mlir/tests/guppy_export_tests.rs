use rstest::{fixture, rstest};
use std::fs;

use hugr_mlir::Result;

#[fixture]
pub fn test_context() -> melior::Context {
    let dr = melior::dialect::DialectRegistry::new();
    let hugr_dh = hugr_mlir::mlir::hugr::get_hugr_dialect_handle();
    hugr_dh.insert_dialect(&dr);
    let ctx = melior::Context::new();
    ctx.append_dialect_registry(&dr);
    ctx.load_all_available_dialects();
    ctx
}

#[rstest]
fn test_guppy_exports(test_context: melior::Context) -> Result<()> {
    use hugr_mlir::hugr_to_mlir::hugr_to_mlir;
    // unsafe { mlir_sys::mlirEnableGlobalDebug(true); }
    insta::glob!("guppy-exports", "*.json", |path| {
        let bytes = fs::read(path).unwrap();
        let ul = melior::ir::Location::new(&test_context, path.to_str().unwrap(), 0, 0);
        let hugr = serde_json::from_slice::<hugr::Hugr>(&bytes).unwrap();
        let mlir_mod = hugr_to_mlir(ul, &hugr).unwrap();
        // println!("{:?}:{}", path, mlir_mod.as_operation());
        assert!(mlir_mod.as_operation().verify());
        insta::assert_display_snapshot!(mlir_mod.as_operation());

        // let hugr2 = mlir_to_hugr(&mlir_mod.as_operation()).unwrap();
        // let mlir_mod2 = hugr_to_mlir(ul, &hugr2).unwrap();
        // assert_eq!(mlir_mod2.as_operation(), mlir_mod.as_operation());
    });
    Ok(())
}
