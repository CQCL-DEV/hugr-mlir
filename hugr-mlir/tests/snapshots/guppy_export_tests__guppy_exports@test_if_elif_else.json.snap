---
source: hugr-mlir/tests/guppy_export_tests.rs
expression: mlir_mod.as_operation()
input_file: hugr-mlir/tests/guppy-exports/test_if_elif_else.json
---
module {
  hugr.module {
    func @foo[](%arg0: !hugr.sum<tuple<>, tuple<>>, %arg1: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> {
      %0 = dfg input extensions [] %arg0, %arg1 : (!hugr.sum<tuple<>, tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> {
      ^bb0(%arg2: !hugr.sum<tuple<>, tuple<>>, %arg3: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
        %1 = cfg %arg2, %arg3 : (!hugr.sum<tuple<>, tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> {
        ^bb0(%arg4: !hugr.sum<tuple<>, tuple<>>, %arg5: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
          %2:2 = dfg input extensions [] %arg4, %arg5 : (!hugr.sum<tuple<>, tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> (!hugr.sum<tuple<>, tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) {
          ^bb0(%arg6: !hugr.sum<tuple<>, tuple<>>, %arg7: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
            output %arg6, %arg7 : !hugr.sum<tuple<>, tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
          }
          switch %2#0, %2#1 : !hugr.sum<tuple<>, tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> ^bb2, ^bb3
        ^bb1(%3: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):  // pred: ^bb4
          output %3 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
        ^bb2(%4: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):  // pred: ^bb0
          %5:2 = dfg input extensions [] %4 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> (!hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) {
          ^bb0(%arg6: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
            const @const_26 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> = 1 : i6
            %15 = load_constant @const_26 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
            %16 = ext_op []"iadd" %arg6, %15 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
            %17 = make_tuple()
            %18 = tag 0 %17 : tuple<> -> <tuple<>>
            output %18, %16 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
          }
          switch %5#0, %5#1 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> ^bb4
        ^bb3(%6: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):  // pred: ^bb0
          %7 = dfg input extensions [] %6 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<>> {
          ^bb0(%arg6: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
            const @const_32 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> = 4 : i6
            %15 = load_constant @const_32 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
            %16 = ext_op []"igt_s" %arg6, %15 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr.sum<tuple<>, tuple<>>
            %17 = make_tuple(%arg6 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>)
            %18 = make_tuple()
            %19 = conditional(%16, %17, %18 : !hugr.sum<tuple<>, tuple<>>, tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<>) -> !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<>> {
            ^bb0(%arg7: tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, %arg8: tuple<>):
              %20 = dfg input extensions [] %arg7, %arg8 : (tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<>) -> !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<>> {
              ^bb0(%arg9: tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, %arg10: tuple<>):
                %21 = tag 0 %arg9 : tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>> -> <tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<>>
                output %21 : !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<>>
              }
              output %20 : !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<>>
            }, {
            ^bb0(%arg7: tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, %arg8: tuple<>):
              %20 = dfg input extensions [] %arg7, %arg8 : (tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<>) -> !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<>> {
              ^bb0(%arg9: tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, %arg10: tuple<>):
                %21 = tag 1 %arg10 : tuple<> -> <tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<>>
                output %21 : !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<>>
              }
              output %20 : !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<>>
            }
            output %19 : !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<>>
          }
          switch %7 : !hugr.sum<tuple<!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>>, tuple<>> ^bb5, ^bb6
        ^bb4(%8: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):  // 2 preds: ^bb2, ^bb7
          %9:2 = dfg input extensions [] %8 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> (!hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) {
          ^bb0(%arg6: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
            %15 = make_tuple()
            %16 = tag 0 %15 : tuple<> -> <tuple<>>
            output %16, %arg6 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
          }
          switch %9#0, %9#1 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> ^bb1
        ^bb5(%10: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):  // pred: ^bb3
          %11:2 = dfg input extensions [] %10 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> (!hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) {
          ^bb0(%arg6: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
            const @const_46 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> = 7 : i6
            %15 = load_constant @const_46 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
            %16 = ext_op []"imul" %arg6, %15 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
            %17 = make_tuple()
            %18 = tag 0 %17 : tuple<> -> <tuple<>>
            output %18, %16 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
          }
          switch %11#0, %11#1 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> ^bb7
        ^bb6:  // pred: ^bb3
          %12:2 = dfg input extensions []  : () -> (!hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) {
            const @const_52 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> = -7 : i6
            %15 = load_constant @const_52 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
            %16 = make_tuple()
            %17 = tag 0 %16 : tuple<> -> <tuple<>>
            output %17, %15 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
          }
          switch %12#0, %12#1 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> ^bb7
        ^bb7(%13: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):  // 2 preds: ^bb5, ^bb6
          %14:2 = dfg input extensions [] %13 : (!hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) -> (!hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>) {
          ^bb0(%arg6: !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>):
            %15 = make_tuple()
            %16 = tag 0 %15 : tuple<> -> <tuple<>>
            output %16, %arg6 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
          }
          switch %14#0, %14#1 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>> ^bb4
        }
        output %1 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
      }
      output %0 : !hugr<opaque "int"["arithmetic.int.types"] <6 : i64>>
    }
  }
}

